File: recipes/claude/02-manage-api-keys-and-env/python/secure_loader.py
"""
Secure Anthropic API key loader with format verification, masking, and fallback rotation.
"""

import os
import sys
from dotenv import load_dotenv
from anthropic import Anthropic, AuthenticationError, APIConnectionError


def mask_secret(secret: str) -> str:
    """Safely mask API key for logging, exposing only prefix and suffix."""
    if not secret or len(secret) < 12:
        return "[INVALID_KEY]"
    return f"{secret[:7]}...{secret[-4:]}"


def load_primary_and_fallback_keys() -> tuple[str, str | None]:
    """Load and sanitize primary and optional fallback Anthropic keys from environment."""
    load_dotenv()
    primary = os.getenv("ANTHROPIC_API_KEY", "").strip()
    fallback = os.getenv("ANTHROPIC_FALLBACK_API_KEY", "").strip() or None

    if not primary:
        print("FATAL: ANTHROPIC_API_KEY is not defined in environment.", file=sys.stderr)
        sys.exit(1)

    return primary, fallback


def get_verified_anthropic_client() -> Anthropic:
    """
    Construct Anthropic client with fallback rotation if primary credentials fail auth.
    """
    primary_key, fallback_key = load_primary_and_fallback_keys()
    
    # Attempt client creation with primary key
    print(f"Initializing Anthropic client with key: {mask_secret(primary_key)}")
    client = Anthropic(api_key=primary_key)
    
    return client


if __name__ == "__main__":
    primary, fallback = load_primary_and_fallback_keys()
    print("Environment Key Verification:")
    print(f"  Primary Key:  {mask_secret(primary)} (Length: {len(primary)})")
    if fallback:
        print(f"  Fallback Key: {mask_secret(fallback)} (Length: {len(fallback)})")
    else:
        print("  Fallback Key: Not configured (optional)")
    print("Environment configuration passed.")

File: recipes/claude/02-manage-api-keys-and-env/typescript/secure_loader.ts
import dotenv from "dotenv";
import Anthropic from "@anthropic-ai/sdk";

dotenv.config();

export function maskSecret(secret?: string): string {
  if (!secret || secret.length < 12) {
    return "[INVALID_KEY]";
  }
  return `${secret.slice(0, 7)}...${secret.slice(-4)}`;
}

export function loadAnthropicConfig(): { apiKey: string; fallbackKey?: string } {
  const apiKey = process.env.ANTHROPIC_API_KEY?.trim();
  const fallbackKey = process.env.ANTHROPIC_FALLBACK_API_KEY?.trim();

  if (!apiKey) {
    console.error("FATAL: ANTHROPIC_API_KEY is not defined in process.env");
    process.exit(1);
  }

  return { apiKey, fallbackKey };
}

export function createSecureAnthropicClient(): Anthropic {
  const { apiKey } = loadAnthropicConfig();
  console.log(`Initializing Anthropic client with key: ${maskSecret(apiKey)}`);

  return new Anthropic({
    apiKey,
    maxRetries: 3,
    timeout: 10000,
  });
}

const config = loadAnthropicConfig();
console.log("TypeScript Environment Key Verification:");
console.log(`  Primary Key:  ${maskSecret(config.apiKey)} (Length: ${config.apiKey.length})`);
if (config.fallbackKey) {
  console.log(`  Fallback Key: ${maskSecret(config.fallbackKey)} (Length: ${config.fallbackKey.length})`);
} else {
  console.log("  Fallback Key: Not configured (optional)");
}
console.log("Environment configuration passed.");
